Controller.cpp

```diff
@@ -523,6 +591,28 @@ namespace Procon {
 		return sendCommand(0x10, buf);
 	}
 
+	Controller::exchangeArray Controller::getSpiData(uint32_t offset, const uint8_t readLen)
+	{
+		std::array<uint8_t, 0x100> buf;
+		while (true)
```

Comment:
I don't like this, since if the connection drops `sendSubcommand` will always return `{}` so this loop will never break.

Reply:
The problem is, that I do not know how often you are supposed to ask for the correct address until you will actually get it. Perhaps we need to dig deeper here.

Reply:
The problem here is not only signal quality but also if the x30 input report is enabled.
If it's enabled, it will parse x30 input reports additionally to x21 reports.

The correct way to mitigate this is to use a thread for packet buffering that reads the id and sends it to the correct parser function. Like what Switch does.

The easiest way is to make sure that the procon uses the x3F input report style, when sending subcmds that you want to parse their reply. 
But you still need to check if you received a x3F reply, because user pressed a button.
And of course add some code for max retries, to avoid infinite loops situations.

Reply:
I tried to enable 0x3F mode first read spi and enable 0x30 mode. But like 50% of the times I start the application it is not working. I wonder if we need to wait for a proper ACK before sending the next command?

Reply:
Normally depends on specific situations.
Pro controller and Joy-Con on a charging grip (both USB) tend to enable x30 mode, if I remember correctly, when you enable some functions. 
In BT mode both of them abide right away to the x03 subcmd.

The first thing to try is to add a wait of around 32ms after enabling x3F mode, before sending any command, to be sure that you will not parse any x30.

But for sure, because there's no any packet director, it's good to have the while (true) code and break out of it when you got the correct ack or after a maximum of retries. 
You are bound to have unwanted packets for the procon and joy-con.